src/main/java/org/apache/pulsar/logstash/outputs/Pulsar.java

```diff
@@ -222,15 +229,25 @@ public void output(final Collection<Event> events) {
                 codec.encode(event, baos);
                 String s = baos.toString();
                 logger.debug("topic is {}, message is {}", eventTopic, s);
-                getProducer(eventTopic).newMessage()
-                        .value(s.getBytes())
-                        .sendAsync();
+                send(eventTopic, s);
             } catch (Exception e) {
                 logger.error("fail to send message", e);
             }
         }
     }
 
+    private void send(String topic, String s){
+        if (enableAsync){
+            getProducer(topic).newMessage()
+                .value(s.getBytes())
+                .sendAsync();
```

Comment:
We still need to add a callback to print a log when failing to publish messages.

Reply:
This send function is only called inside a try/catch, where it logs "fail to send message" with the exception. See line 225-235.

Reply:
It won't catch the exception thrown from `sendAsync`. We need to use `CompletableFuture.exceptionally` to catch the it. Please see: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#exceptionally-java.util.function.Function-

Reply:
Okay.. Am I reading this right? 

CompletableFuture.exceptionally description:

> Returns a new CompletableFuture that is completed when this CompletableFuture completes, with the result of the given function of the exception triggering this CompletableFuture's completion when it completes exceptionally; otherwise, if this CompletableFuture completes normally, then the returned CompletableFuture also completes normally with the same value.

This doesn't throw a custom Exception that inherits from the class Exception as usual? I don't think I understand what this is describing. I'm trying to decipher it, but does it mean that it returns the function that triggered an Exception rather than actually throwing the Exception. In that case I guess I couldn't try/catch? Or does it mean that I am supposed to just provide a function that is throwable, so when CompletableFuture throws, it triggers my function with where it failed?

Reply:
Sorry for the late reply.

The `sendAsync` will not throw any exceptions. The producer already catches all the exceptions and places them into the `Future`. You could check the code [here](https://github.com/apache/pulsar/blob/82237d3684fe506bcb6426b3b23f413422e6e4fb/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBase.java#L67-L71).

Therefore if there are any exceptions thrown in the send operation, the function `fn` in the `exceptionally` will be called. And I think we need to print the error log in that function.

> Or does it mean that I am supposed to just provide a function that is throwable, so when CompletableFuture throws, it triggers my function with where it failed?

That's correct.

However,  you still need to use try/catch here, because we couldn't guarantee that this code won't throw the exception:
```
getProducer(topic).newMessage()
                .value(s.getBytes())
                .sendAsync();
```
For instance, if the `s` is null.